api/src/main/java/org/pmiops/workbench/api/UserMetricsController.java

```diff
@@ -193,31 +213,84 @@ public void setDistinctWorkspaceLimit(int limit) {
     return ResponseEntity.ok(recentResponse);
   }
 
-  private Boolean foundBlobIdsContainsUserRecentResource(
-      Set<BlobId> foundNotebooks, DbUserRecentResource urr) {
-    return Optional.ofNullable(urr.getNotebookName())
-        .flatMap(this::uriToBlobId)
-        .map(foundNotebooks::contains)
-        .orElse(true);
+  private Boolean foundBlobIdsContainsUserRecentlyModifiedResource(
+      Set<BlobId> foundNotebooks, DbUserRecentlyModifiedResource urr) {
+    if (urr.getResourceType()
+        == DbUserRecentlyModifiedResource.DbUserRecentlyModifiedResourceType.NOTEBOOK) {
+      return Optional.ofNullable(urr.getResourceId())
+          .flatMap(this::uriToBlobId)
+          .map(foundNotebooks::contains)
+          .orElse(true);
+    }
+    return true;
   }
 
   @VisibleForTesting
-  public boolean hasValidBlobIdIfNotebookNamePresent(DbUserRecentResource dbUserRecentResource) {
-    return Optional.ofNullable(dbUserRecentResource.getNotebookName())
-        .map(name -> uriToBlobId(name).isPresent())
-        .orElse(true);
+  public boolean hasValidBlobIdIfNotebookNamePresent(
+      DbUserRecentlyModifiedResource dbUserRecentResource) {
+    if (dbUserRecentResource.getResourceType()
+        == DbUserRecentlyModifiedResource.DbUserRecentlyModifiedResourceType.NOTEBOOK) {
+      return Optional.ofNullable(dbUserRecentResource.getResourceId())
+          .map(name -> uriToBlobId(name).isPresent())
+          .orElse(true);
+    }
+    return true;
   }
 
+  /**
+   * Build recent resource object by grabbing the DB object (cohort/conceptSet or dataSet) depending
+   * upon the DbUserRecentlyModifiedResource.ResourceType
+   *
+   * @param idToDbWorkspace
+   * @param idToFcWorkspaceResponse
+   * @param dbUserRecentlyModifiedResource
+   * @return WorkspaceResource
+   */
   private WorkspaceResource buildRecentResource(
       Map<Long, DbWorkspace> idToDbWorkspace,
       Map<Long, FirecloudWorkspaceResponse> idToFcWorkspaceResponse,
-      DbUserRecentResource dbUserRecentResource) {
+      DbUserRecentlyModifiedResource dbUserRecentlyModifiedResource) {
+
+    final long workspaceId = dbUserRecentlyModifiedResource.getWorkspaceId();
```

Comment:
Would it be possible to push this logic down to the mapper?

Reply:
(the whole switch I mean, and perhaps more)

Reply:
Done